Symptoms reporting

ivnsch opened the pull request:
Status: "It compiles!" and a couple of Unit Tests.

Commit:
Implement public report <-> memo

duskoo commented on src/memo.rs:
```diff
@@ -0,0 +1,432 @@
+struct Memo {
+  bytes: Vec<u8>
+}
+#[derive(Debug, PartialEq, Clone)]
```
@i-schuetz Why not use Eq instead of PartialEq? 
Is reflexivity a problem?

ivnsch replied:
I think I used it because it's what the compiler requires (e.g. to use `==`), but yeah, `Eq` is semantically correct.

ivnsch replied:
(Will also fix in a follow up PR)

duskoo commented on src/memo.rs:
```diff
@@ -0,0 +1,432 @@
+  bytes: Vec<u8>
+}
+#[derive(Debug, PartialEq, Clone)]
+enum UserInput<T> {
```
This is renamed optional : https://doc.rust-lang.org/std/option/
We want to point out that data source is user input... ok

ivnsch replied:
Yup

duskoo commented on src/memo.rs:
```diff
@@ -0,0 +1,432 @@
+  const FEVER_SEVERITY_MAPPER: FeverSeverityMapper = FeverSeverityMapper {};
+  const BOOLEAN_MAPPER: BoolMapper = BoolMapper {};
+
+  fn extract<T>(bits: &Vec<bool>, mapper: &dyn BitMapper<T>, start: usize) -> ExtractResult<T> {
```
usize can vary depending on device: https://doc.rust-lang.org/std/primitive.usize.html . Better to use type with fixed size.

ivnsch replied:
`usize` is required by the compiler for [indexing](https://doc.rust-lang.org/1.22.1/std/ops/trait.Index.html). 32 bits `usize` allows to index 4,294,967,295 values. I don't know what happens if trying to index a number greater than that on a 32bit architecture, but it appears safe, see
https://users.rust-lang.org/t/i32-vs-isize-u32-vs-usize/22657
https://stackoverflow.com/a/29592369/930450

(For completeness, for the memo we currently have to index 2056 bits).

duskoo commented on src/memo.rs:
```diff
@@ -0,0 +1,432 @@
+
+  fn extract<T>(bits: &Vec<bool>, mapper: &dyn BitMapper<T>, start: usize) -> ExtractResult<T> {
+    let end = mapper.bit_count() + start;
+    let sub_bits: Vec<bool> = bits[start..end].try_into().expect("msg");
```
"msg" -> "Conversion to MemoMapperImpl failed"

Commit:
Split in modules

duskoo commented on src/memo.rs:
```diff
@@ -0,0 +1,432 @@
+  }
+}
+
+struct BitList {
```
Its not really a list. Why not call it BitVector (MemoBitVector if there is a name collision).

Commit:
Implement symptoms processing and keys management

- FFI function to submit symptoms
- Inputs form and mapping to public report
- Storage of generated RAK and TCK (TCK not used for now, as TCNs not generated in lib yet) in preferences

duskoo commented on src/composition_root.rs:
```diff
@@ -1,28 +1,49 @@
+  PreferencesImpl, TcnDaoImpl, TcnMatcherImpl, TcnApiImpl, 
+  SymptomInputsSubmitterImpl<MemoMapperImpl, TcnKeysImpl<PreferencesImpl>, TcnApiImpl>
+> {
+  // FIXME pass the same instances / references
```
https://doc.rust-lang.org/book/ch05-01-defining-structs.html#ownership-of-struct-data

ivnsch replied:
I'm merging it like this for now. Attempted to pass references / use lifetimes, but it fails with the `Preferences`, because they use `RwLock` and this uses unsafe code internally. Will fix in a follow up PR. In case you want to give it a look too, here's a self contained snippet:
```rust
struct MyPrefs<'a> {
  field1: RwLock<&'a str>
}

struct Dependency<'a> {
  field1: &'a MyPrefs<'a>
}

struct DependencyGraph<'a> {
  dep: Dependency<'a>,
}

fn create_dep_graph() -> DependencyGraph<'static> {
  let my_prefs = &MyPrefs { field1: RwLock::new("") };
  let dep = Dependency { field1: my_prefs };
  DependencyGraph { dep }
}
```

i-schuetz replied:
And here the snippet again using `RwLock`
```rust
struct MyPrefs<'a> {
  field1: RwLock<&'a str>
}

struct Dependency2<'a> {
  field1: &'a MyPrefs<'a>
}

struct DependencyGraph<'a> {
  dep2: Dependency2<'a>,
}

fn create_dep_graph() -> DependencyGraph<'static> {
  let my_prefs = &MyPrefs { field1: RwLock::new("") };
  let dep2 = Dependency2 { field1: my_prefs };
  DependencyGraph { dep2 }
}
```

duskoo commented on src/reporting/bit_list.rs:
```diff
@@ -0,0 +1,74 @@
+impl BitList {
+
+  pub fn concat(&self, bit_list: BitList) -> BitList {
+    BitList { bits: self.bits.clone().into_iter().chain(bit_list.bits.into_iter()).collect() }
```
Can extend be used here: https://doc.rust-lang.org/std/iter/trait.Extend.html
(instead of chain/collect)

ivnsch replied:
`extend` appears to mutate the vector. Unless we've good reasons I'd prefer to keep immutable/flow style for now. I'd be interested in something to make this shorter though. If there's nothing out of the box, writing our own extensions. We have to investigate.

ivnsch replied:
P.S. Given Rust's safety around mutability, I'm open to use it. I find immutability easier to reason about / write in general: e.g. `return old_vec + vec` instead of `old_vec.add(vec); return old_vec`. Maybe we also can write extensions that use mutability internally, with an immutable interface.

duskoo commented on src/reporting/bit_list.rs:
```diff
@@ -0,0 +1,74 @@
+use std::{u64, convert::TryInto};
+
+pub struct BitList {
```
Unusual name for a vector of bool

Commit:
Rename BitList in BitVector
